### PR TITLE
Added clipBehavior to Overlay, Flow, AnimatedSize and AndroidView

### LINKED ISSUE #63246: Add clipBehavior to Overlay, Flow, AnimatedSize, and AndroidView
Thanks for https://github.com/flutter/flutter/pull/63147#issuecomment-670435675 to point this out.

### PR DESCRIPTION
**Pull request by thecalamiity**
## Description

Added clipBehavior to Overlay, Flow, AnimatedSize and AndroidView.

/cc @liyuqian 

## Related Issues

Closes [#63246]

## Tests

Added tests to check if the mentioned widgets can set and update their clipBehaviors.

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[#63246]: https://github.com/flutter/flutter/issues/63246

### CODE REVIEW

File: packages/flutter/lib/src/rendering/animated_size.dart [hunk 1]
```diff
@@ -139,6 +142,20 @@ class RenderAnimatedSize extends RenderAligningShiftedBox {
     _animation.curve = value;
   }
 
+  /// {@macro flutter.widgets.Clip}
+  ///
+  /// Defaults to [Clip.none], and must not be null.
+  Clip get clipBehavior => _clipBehavior;
+  Clip _clipBehavior = Clip.none;
```
Comment (liyuqian) [inline packages/flutter/lib/src/rendering/animated_size.dart:149]: ditto.

File: packages/flutter/lib/src/rendering/flow.dart [hunk 2]
```diff
@@ -181,8 +181,11 @@ class RenderFlow extends RenderBox
   RenderFlow({
     List<RenderBox>? children,
     required FlowDelegate delegate,
+    Clip clipBehavior = Clip.none,
```
Comment (liyuqian) [inline packages/flutter/lib/src/rendering/flow.dart:184]: ditto.

File: packages/flutter/lib/src/rendering/flow.dart [hunk 3]
```diff
@@ -221,6 +224,20 @@ class RenderFlow extends RenderBox
     }
   }
 
+  /// {@macro flutter.widgets.Clip}
+  ///
+  /// Defaults to [Clip.none], and must not be null.
+  Clip get clipBehavior => _clipBehavior;
+  Clip _clipBehavior = Clip.none;
```
Comment (liyuqian) [inline packages/flutter/lib/src/rendering/flow.dart:231]: ditto

File: packages/flutter/lib/src/rendering/platform_view.dart [hunk 4]
```diff
@@ -84,10 +84,13 @@ class RenderAndroidView extends RenderBox with _PlatformViewGestureMixin {
     required AndroidViewController viewController,
     required PlatformViewHitTestBehavior hitTestBehavior,
     required Set<Factory<OneSequenceGestureRecognizer>> gestureRecognizers,
+    Clip clipBehavior = Clip.none,
```
Comment (liyuqian) [inline packages/flutter/lib/src/rendering/platform_view.dart:87]: ditto

File: packages/flutter/lib/src/rendering/platform_view.dart [hunk 5]
```diff
@@ -116,6 +119,20 @@ class RenderAndroidView extends RenderBox with _PlatformViewGestureMixin {
     _viewController.addOnPlatformViewCreatedListener(_onPlatformViewCreated);
   }
 
+  /// {@macro flutter.widgets.Clip}
+  ///
+  /// Defaults to [Clip.none], and must not be null.
+  Clip get clipBehavior => _clipBehavior;
+  Clip _clipBehavior = Clip.none;
```
Comment (liyuqian) [inline packages/flutter/lib/src/rendering/platform_view.dart:126]: ditto

File: packages/flutter/lib/src/widgets/basic.dart [hunk 6]
```diff
@@ -5091,7 +5091,9 @@ class Flow extends MultiChildRenderObjectWidget {
     Key? key,
     required this.delegate,
     List<Widget> children = const <Widget>[],
+    this.clipBehavior = Clip.none,
```
Comment (liyuqian) [inline packages/flutter/lib/src/widgets/basic.dart:5094]: ditto

File: packages/flutter/lib/src/widgets/basic.dart [hunk 7]
```diff
@@ -5106,18 +5108,26 @@ class Flow extends MultiChildRenderObjectWidget {
...
```
Comment (liyuqian) [inline packages/flutter/lib/src/widgets/basic.dart:5111]: ditto

File: packages/flutter/lib/src/widgets/overlay.dart [hunk 8]
```diff
@@ -214,7 +214,9 @@ class Overlay extends StatefulWidget {
   const Overlay({
     Key? key,
     this.initialEntries = const <OverlayEntry>[],
+    this.clipBehavior = Clip.none,
```
Comment (liyuqian) [inline packages/flutter/lib/src/widgets/overlay.dart:217]: ditto

File: packages/flutter/lib/src/widgets/overlay.dart [hunk 9]
```diff
@@ -232,6 +234,11 @@ class Overlay extends StatefulWidget {
   /// To remove an entry from an [Overlay], use [OverlayEntry.remove].
   final List<OverlayEntry> initialEntries;
 
+  /// {@macro flutter.widgets.Clip}
+  ///
+  /// Defaults to [Clip.none], and must not be null.
```
Comment (liyuqian) [inline packages/flutter/lib/src/widgets/overlay.dart:239]: remember to change documentation too when the default is changed to `Clip.hardEdge`.

File: packages/flutter/lib/src/widgets/overlay.dart [hunk 10]
```diff
@@ -491,15 +499,19 @@ class _Theatre extends MultiChildRenderObjectWidget {
   _Theatre({
     Key? key,
     this.skipCount = 0,
+    this.clipBehavior = Clip.none,
```
Comment (liyuqian) [inline packages/flutter/lib/src/widgets/overlay.dart:502]: ditto

File: packages/flutter/lib/src/widgets/overlay.dart [hunk 11]
```diff
@@ -546,11 +560,14 @@ class _RenderTheatre extends RenderBox with ContainerRenderObjectMixin<RenderBox
     List<RenderBox>? children,
     required TextDirection textDirection,
     int skipCount = 0,
+    Clip clipBehavior = Clip.none,
```
Comment (liyuqian) [inline packages/flutter/lib/src/widgets/overlay.dart:563]: ditto

File: packages/flutter/lib/src/widgets/overlay.dart [hunk 12]
```diff
@@ -594,6 +611,20 @@ class _RenderTheatre extends RenderBox with ContainerRenderObjectMixin<RenderBox
     }
   }
 
+  /// {@macro flutter.widgets.Clip}
+  ///
+  /// Defaults to [Clip.none], and must not be null.
+  Clip get clipBehavior => _clipBehavior;
+  Clip _clipBehavior = Clip.none;
```
Comment (liyuqian) [inline packages/flutter/lib/src/widgets/overlay.dart:618]: ditto

**Review by liyuqian — COMMENTED** [review]
Thanks for writing this patch! It mostly looks good, but let's avoid the breaking change, a very lengthy process, by setting default clip to hardEdge for now. I think that would also fix some widget test failures with clip assertions.

**Comment by TheBiirb** [comment]
@liyuqian Did the same to AndroidView and AnimatedSize because some tests were failing

**Review by liyuqian — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/rendering/flow.dart [hunk 13]
```diff
@@ -365,7 +382,11 @@ class RenderFlow extends RenderBox
 
   @override
   void paint(PaintingContext context, Offset offset) {
-    context.pushClipRect(needsCompositing, offset, Offset.zero & size, _paintWithDelegate);
+    if(clipBehavior == Clip.none) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/rendering/flow.dart:385]: nit: space after `if` and before `(`.

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/rendering/flow.dart (hunk 13 above)
Comment (TheBiirb) [reply packages/flutter/lib/src/rendering/flow.dart]: Done :D

File: packages/flutter/lib/src/rendering/platform_view.dart [hunk 14]
```diff
@@ -189,8 +206,8 @@ class RenderAndroidView extends RenderBox with _PlatformViewGestureMixin {
 
     // Clip the texture if it's going to paint out of the bounds of the renter box
     // (see comment in _paintTexture for an explanation of when this happens).
-    if (size.width < _currentAndroidViewSize.width || size.height < _currentAndroidViewSize.height) {
-      context.pushClipRect(true, offset, offset & size, _paintTexture);
+    if (size.width < _currentAndroidViewSize.width || size.height < _currentAndroidViewSize.height && clipBehavior != Clip.none) {
```
Comment (flar) [inline packages/flutter/lib/src/rendering/platform_view.dart:209]: `&&` has higher precedence than `||`

I have no idea what the logical flow was intended to be here, but it looks wrong.

**Review by flar — COMMENTED** [review]
This is already merged, but I think that line of code needs to be reexamined. I will submit an issue to double check it.

File: packages/flutter/lib/src/rendering/platform_view.dart (hunk 14 above)
Comment (goderbauer) [reply packages/flutter/lib/src/rendering/platform_view.dart:209]: That does look wrong. Wanna send a PR to fix it?

Comment (flar) [reply packages/flutter/lib/src/rendering/platform_view.dart:209]: See https://github.com/flutter/flutter/issues/67343

Comment (liyuqian) [reply packages/flutter/lib/src/rendering/platform_view.dart:209]: Agree it's wrong. Thanks for finding it!

Comment (TheBiirb) [reply packages/flutter/lib/src/rendering/platform_view.dart:209]: So sorry for this guys 😞

**Pull request merged**